sycl/plugins/level_zero/pi_level0.cpp

```diff
@@ -0,0 +1,3732 @@
+//===----------- pi_level0.cpp - Level Zero Plugin--------------------------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+/// \file pi_level0.cpp
+/// Implementation of Level Zero Plugin.
+///
+/// \ingroup sycl_pi_level0
+
+#include "pi_level0.hpp"
+#include <cstdarg>
+#include <cstdio>
+#include <cstring>
+#include <map>
+#include <memory>
+#include <string>
+#include <thread>
+
+#include <level_zero/zet_api.h>
+
+// Controls L0 calls serialization to w/a L0 driver being not MT ready.
+// Recognized values (can be used as a bit mask):
+enum {
+  ZeSerializeNone =
+      0, // no locking or blocking (except when SYCL RT requested blocking)
+  ZeSerializeLock = 1, // locking around each ZE_CALL
+  ZeSerializeBlock =
+      2, // blocking ZE calls, where supported (usually in enqueue commands)
+};
+static pi_uint32 ZeSerialize = 0;
+
+// This class encapsulates actions taken along with a call to L0 API.
+class ZeCall {
+private:
+  // The global mutex that is used for total serialization of L0 calls.
+  static std::mutex GlobalLock;
+
+public:
+  ZeCall() {
+    if ((ZeSerialize & ZeSerializeLock) != 0) {
+      GlobalLock.lock();
+    }
+  }
+  ~ZeCall() {
+    if ((ZeSerialize & ZeSerializeLock) != 0) {
+      GlobalLock.unlock();
+    }
+  }
+
+  static ze_result_t check(ze_result_t ZeResult, const char *CallStr,
+                           bool TraceError = true);
+
+  // The non-static version just calls static one.
+  ze_result_t checkThis(ze_result_t ZeResult, const char *CallStr,
+                        bool TraceError = true) {
+    return ZeCall::check(ZeResult, CallStr, TraceError);
+  }
+};
```

Comment:
Does this class wrap-around calls to L0 backend API?
If so, I think this class should have some method like `act` of `doCall` or smth like this to perform call to L0 backend API.

Does this class have any derivatives?
If so, there's a possible memory leak if there's an exception thrown from constructor as no destructor will be called. To overcome this I suggest only initializing some `std::unique_lock` with `std::defer_lock` and perform actual locking elsewhere prior to actual call to L0 backend API.

Reply:
Thanks, renamed to more meaningful doCall and removed unnecessary static method. 
This class doesn't have any derivatives.